content/blog/2025/11/nuttx-external-lib/index.md

```diff
@@ -0,0 +1,196 @@
+---
+title: "Integrating External Libraries into NuttX Applications"
+date: 2025-11-17
+tags: ["NuttX", "Apache", "ESP32", "ESP32-C6", "RISC-V", "cross-compilation", "static-library"]
+showAuthor: false
+authors:
+    - "filipe-cavalcanti"
+summary: "This guide demonstrates how to integrate external libraries into NuttX applications using static libraries and cross-compilation. Learn how to build a library on x86, integrate it into the NuttX simulation environment, and cross-compile for RISC-V targets like the ESP32-C6, all without moving your codebase into the NuttX directory structure."
+---
+
+## Introduction
+
+When moving your application to NuttX, you often need to add your existing software stack to the NuttX image. This software may run on a different RTOS or even on an x86 environment, but sometimes it must run on multiple target devices. This article shows how to build NuttX with your custom application without moving your entire stack to the NuttX directory. You use a static library and cross-compilation to achieve this.
+
+This article has three parts: building a sample library on x86, adding it to the NuttX simulation environment, and finally, cross-compiling to RISC-V and running it on the ESP32-C6.
+
+## Using an example library
+
+As an example, use an application that converts a hexadecimal color string to RGB. The reference project is available in [this repository](https://github.com/fdcavalcanti/hex-converter) and has the following structure:
+
+```
+hex-converter/
+├── src/
+│   ├── hex_to_rgb.h
+│   └── hex_to_rgb.c
+├── main.c
+├── test.c
+├── Makefile
+└── README.md
+```
+
+The `hex-converter` library exposes one single function called `hex_to_rgb`. The user provides a pointer to a string representing the hex color and a pointer
+to an array where the R, G, and B values are copied. It is a simple application but very useful as an example.
+
+Clone the repository and build it according to the steps in the README file to produce the static library. To confirm that everything works, run the provided `main` example program. This program accepts a hexadecimal color string as input.
+
+```bash
+$ ./main "#1A2B3C"
+Input: #1A2B3C
+RGB: (26, 43, 60)
+```
+
+At this point, the directory should contain a static library called `libhex_to_rgb.a` that will be added to the NuttX build system.
+
+## Testing on NuttX Simulation 
+
+As a user, I want to use that library in my application. The first solution might be to copy the entire hex-converter repository to the NuttX application directory and 
```

Comment:
"As a user, I want to use that library in my application." -> "As a user, you might want to use this library in your application."